fiftyone/core/dataset.py

```diff
@@ -4572,6 +4620,61 @@ def _reload_docs(self, hard=False):
         if self.media_type == fom.VIDEO:
             fofr.Frame._reload_docs(self._frame_collection_name, hard=hard)
 
+    def _save_samples(self):
+        if self._sample_save_ops and len(self._sample_save_ops) > 0:
+            try:
+                foo.bulk_write(
+                    self._sample_save_ops,
+                    self._sample_collection,
+                    ordered=False,
+                )
+            except BulkWriteError as bwe:
+                msg = bwe.details["writeErrors"][0]["errmsg"]
+                raise ValueError(msg) from bwe
+            finally:
+                self._sample_save_ops.clear()
+
+    def _start_sample_save_batching(self):
+        if (
+            self._sample_save_batch_size is not None
+            and self._sample_save_ops is None
+        ):
+            self._sample_save_ops = []
+
+    def _stop_sample_save_batching(self):
+        self._save_samples()
+        self._sample_save_batch_size = None
+        self._sample_save_ops = None
+
+    def save_context(self, batch_size):
+        """Puts the dataset in btach save mode when used as a context manager.
+
+        Args:
+            batch_size: number of `sample.save()` to aggregate into
+                one bulk call
+
+        Returns:
+            :class:`fiftyone.core.dataset.Dataset`
+        """
+        self._sample_save_batch_size = batch_size
+        return self
+
+    def batch_save_sample(self, sample):
```

Comment:
Suggestion: make this a private method because we don't intend *users* to directly use it.

Feel free to object to this if you want, but we currently use private methods for everything that a user won't call, including methods that are called by other classes, like this one here.

Reply:
There is a bit of code smell to me, other classes should respect public/private methods (even if we wrote them). While maybe not too bad in this case, following this pattern creates a tight coupling between classes that should be solved by abstraction layers.

Reply:
Yeah some refactoring would definitely be required to satisfy both of these design constraints simultaneously:
(a) All public methods are intended to be user-callable
(b) Classes don't call other classes' private methods

Currently, (a) is achieved at the expense of (b) in some cases.

Reply:
Fair enough. I can make it private for consistency.